src/effects/star_spirits_energy.c

```diff
@@ -23,4 +114,63 @@ void star_spirits_energy_render(EffectInstance* effect) {
     retTask->renderMode |= RENDER_TASK_FLAG_2;
 }
 
-INCLUDE_ASM(s32, "effects/star_spirits_energy", star_spirits_energy_appendGfx);
+void star_spirits_energy_appendGfx(void* effect) {
+    StarSpiritsEnergyFXData* data = ((EffectInstance*)effect)->data.starSpiritsEnergy;
+    s32 unk_04 = data->unk_04;
+    s32 alpha;
+    Matrix4f sp20;
+    s32 i;
+
+    gDPPipeSync(gMasterGfxPos++);
+    gSPSegment(gMasterGfxPos++, 0x09, VIRTUAL_TO_PHYSICAL(((EffectInstance*)effect)->graphics->data));
+
+    shim_guPositionF(sp20, 0.0f, -gCameras[gCurrentCameraID].currentYaw, 0.0f, data->unk_3C, data->unk_08, data->unk_0C, data->unk_10);
+    shim_guMtxF2L(sp20, &gDisplayContext->matrixStack[gMatrixListPos]);
+
+    gSPMatrix(gMasterGfxPos++, &gDisplayContext->matrixStack[gMatrixListPos++], G_MTX_PUSH | G_MTX_LOAD | G_MTX_MODELVIEW);
+    gSPDisplayList(gMasterGfxPos++, D_E0122BA4[0]);
+    gDPSetEnvColor(gMasterGfxPos++, data->unk_2C, data->unk_30, data->unk_34, data->unk_38);
+
+    for (i = 0; i < 4; i++) {
+        if (unk_04 != 1 || i != 2) {
+            if (data->unk_40[i] != 0.0f && data->unk_50[i] != 0) {
+                shim_guPositionF(sp20, 0.0f, 0.0f, 0.0f, data->unk_40[i], 0.0f, 0.0f, 0.0f);
+                shim_guMtxF2L(sp20, &gDisplayContext->matrixStack[gMatrixListPos]);
+
+                gSPMatrix(gMasterGfxPos++, &gDisplayContext->matrixStack[gMatrixListPos++], G_MTX_PUSH | G_MTX_MUL | G_MTX_MODELVIEW);
+                gDPSetPrimColor(gMasterGfxPos++, 0, 0, data->unk_1C, data->unk_20, data->unk_24, data->unk_50[i]);
+                gSPDisplayList(gMasterGfxPos++, D_E0122B90[i]);
+                gSPPopMatrix(gMasterGfxPos++, G_MTX_MODELVIEW);
+            }
+        }
+    }
+
+    gSPPopMatrix(gMasterGfxPos++, G_MTX_MODELVIEW);
+
+    if (unk_04 == 2 || unk_04 == 3) {
+        alpha = 0;
+
+        if (data->unk_70 == 1 || data->unk_70 == 100) {
+            if (data->unk_74 > 2) {
```

Comment:
it'd be nice to && this and the following condition together, since they're both checking if the thing is within a certain range. also applies in the second block below with 0 and 9

Reply:
Putting the two together wasn't working, but it seems it works if I include the outer condition as well:
```c
(data->unk_70 == 1 || data->unk_70 == 100) && data->unk_74 > 2 && data->unk_74 < 11
```